Set of performance improvements

pkozlowski-opensource opened the pull request.

pkozlowski-opensource committed:
perf(ivy): avoid repeated tView.firstTemplatePass writes

pkozlowski-opensource committed:
refactor(ivy): rewrite refreshDynamicEmbeddedViews for clarity

pkozlowski-opensource committed:
perf(ivy): only refresh child components if those are defined in a given view

pkozlowski-opensource committed:
refactor(ivy): remove unnecessary call to setPreviousOrParentTNode

setPreviousOrParentTNode is set in enterView so no need to reset it
just before entering a view.

pkozlowski-opensource committed:
perf(ivy): avoid creating holey LView Arrays

pkozlowski-opensource committed:
perf(ivy): call refreshContentQueries only when there are content queries defined

mhevery commented on packages/core/src/render3/PERF_NOTES.md:
```diff
@@ -22,6 +22,18 @@ Great reads:
 
  See benchmark [here](https://jsperf.com/mono-vs-megamorphic-property-access).
 
+## Packed vs. holey Array
```
thanks for adding this!

mhevery commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -644,11 +642,18 @@ export function createTView(
+  const blueprint = ngDevMode ? new LViewBlueprint !(initialViewLength) :
+                                Array.apply(null, Array(initialViewLength));
+
+  blueprint.fill(null, 0, bindingStartIndex).fill(NO_CHANGE, bindingStartIndex) as LView;
```
I think there is much better way to do this.
```
const blueprint = ngDevMode ? new LViewBlueprint !() : [];
blueprint.fill(null, 0, bindingStartIndex).fill(NO_CHANGE, bindingStartIndex, initialViewLength) as LView;
```

pkozlowski-opensource replied:
This won't work as `.fill` doesn't write past the array length. I've pushed an alternative (if we don't want to use the "apply" trick):

```
 const blueprint = ngDevMode ? new LViewBlueprint !() : [];

  for (let i = 0; i < initialViewLength; i++) {
    blueprint.push(i < bindingStartIndex ? null : NO_CHANGE);
  }
```

mhevery replied:
Really? OK.

pkozlowski-opensource committed:
perf(ivy): stricter null checks

pkozlowski-opensource committed:
perf(ivy): limit creationMode checks

AndrewKushnir commented:
Adding "cleanup" label to indicate that the size limit needs an update (as per discussion due the team sync), so that `test_aio_local_ivy` CircleCI task passes.

pkozlowski-opensource committed:
build(docs-infra): update main bundle size

PR #31839 contains perf and code cleanup changes that add 48 bytes to the
main bundle size. Perf gains and code clarity justify this increase.

Unfortunately the size verification job is rejecting this PR as it reports
size gains from a fixed size and not relative increase of size from a particular PR.

It was decided during the internal team discussion to bump up size limits to
correctly reflect current state of the master and increase from this PR.

pkozlowski-opensource commented:
@AndrewKushnir I've updated the size limit, as discussed. I did it in a separate commit and update to the same value as in the @crisbeto PR #31839 so we are not generating conflicts (or can drop the commit in question if needed).

petebacondarwin reviewed (approved):
```
perf(aio): update main bundle size
```

This does not seem like the correct commit message.
- the subject we use is now docs-infra (aio is deprecated as a subject)
- should it be type perf? I would expect build or ci or even test

Also it would be really helpful for future readers if you explain in the commit message why this has been done (and mention the discussion).

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:260741330:BASE:260741366:1564506967748:38b9304)
- [Ivy Presubmit](http://test/OCL:260741330:BASE:260741797:1564507067670:27ed8ec3)
